usr/src/uts/common/fs/zfs/metaslab.c

```diff
@@ -2854,9 +2933,11 @@ metaslab_group_alloc_normal(metaslab_group_t *mg, zio_alloc_list_t *zal,
 		 * We have just failed an allocation attempt, check
 		 * that metaslab_should_allocate() agrees. Otherwise,
 		 * we may end up in an infinite loop retrying the same
-		 * metaslab.
+		 * metaslab. Keep in mind, it might have happened due
+		 * to an ongoing trim.
```

Comment:
Worst case, we might loop around for a little while, but it shouldn't be infinite. If the lack of allocation success is due to all metaslabs with usable space trimming, then we'll just loop around at least one of them is done with the trim. I suppose the elegance of the code isn't stellar, but we've so far never run into any issues in the field with this logic. If absolutely necessary, we can add some sort of special abort into the loop which checks if we've tried all metaslabs and all were unavailable due to either lack of space or a trim conflict - in which case we'd break out if there were other vdevs in the pool which could satisfy the allocation.